File: Java/Delete-Digits.java
public class Solution {
    /**
     *@param A: A positive integer which has N digits, A is a string.
     *@param k: Remove k digits.
     *@return: A string
     */
    public String DeleteDigits(String A, int k) {
        // write your code here
        if (A.length() == k) return "";
        Stack<Character> stk = new Stack<Character>();
        int popcount = 0;
        for (int i = 0; i < A.length(); i ++) {
            char c = A.charAt(i);
            if (stk.isEmpty()) {
                stk.push(c);
            } else if (Integer.valueOf(c) >= Integer.valueOf(stk.peek())) {
                stk.push(c);
            } else {
                if (popcount < k) {
                    stk.pop();
                    popcount ++;
                    i --;
                } else {
                    stk.push(c);
                }
            }
        }
        while (popcount < k) {
            stk.pop();
            popcount ++;
        }
        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < stk.size(); i ++) {
            sb.append(stk.get(i));
        }
        String res = sb.toString();
        while (res.length() > 1 && res.charAt(0) == '0') {
            res = res.substring(1);
        }
        return res;
    }
}

File: Cpp/Delete-Digits.cpp
class Solution {
public:
    /**
     *@param A: A positive integer which has N digits, A is a string.
     *@param k: Remove k digits.
     *@return: A string
     */
    string DeleteDigits(string A, int k) {
        // wirte your code here
        stack <int> st;
        int popcount = 0;
        for (int i = 0; i < A.size(); i ++) {
            int num = A[i] - '0';
            if (st.empty()) {
                st.push(num);
            } else if (num >= st.top()) {
                st.push(num);
            } else {
                if (popcount < k) {
                    st.pop();
                    i--;
                    popcount ++;
                } else {
                    st.push(num);
                }
            }
        }
        while (popcount < k) {
            popcount ++;
            st.pop();
        }
        string str = "";
        while (!st.empty()) {
            char ch = '0' + st.top();
            str += ch;
            st.pop();
        }
        reverse(str.begin(), str.end());
        while (str.size() > 1 && str[0] == '0') {
            str = str.substr(1, str.size()-1);
        }
        return str;
    }
};
